server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java

```diff
@@ -81,73 +90,116 @@ protected ClusterBlockException checkBlock(GetSnapshotsRequest request, ClusterS
     protected void masterOperation(final GetSnapshotsRequest request, final ClusterState state,
                                    final ActionListener<GetSnapshotsResponse> listener) {
-        try {
-            final String repository = request.repository();
-            final Map<String, SnapshotId> allSnapshotIds = new HashMap<>();
-            final List<SnapshotInfo> currentSnapshots = new ArrayList<>();
-            for (SnapshotInfo snapshotInfo : snapshotsService.currentSnapshots(repository)) {
-                SnapshotId snapshotId = snapshotInfo.snapshotId();
-                allSnapshotIds.put(snapshotId.getName(), snapshotId);
-                currentSnapshots.add(snapshotInfo);
-            }
+        final String[] repositories = request.repositories();
+        transportService.sendRequest(transportService.getLocalNode(), GetRepositoriesAction.NAME,
+                new GetRepositoriesRequest(repositories),
+                new ActionListenerResponseHandler<>(
+                        ActionListener.wrap(
+                                response ->
+                                        // switch to GENERIC thread pool because it might be long running operation
+                                        threadPool.executor(ThreadPool.Names.GENERIC).execute(
+                                                () -> getMultipleReposSnapshotInfo(response.repositories(), request.snapshots(),
+                                                        request.ignoreUnavailable(), request.verbose(), listener)),
+                                listener::onFailure),
+                        GetRepositoriesResponse::new));
+    }
+
+    private void getMultipleReposSnapshotInfo(List<RepositoryMetaData> repos, String[] snapshots, boolean ignoreUnavailable,
+                                              boolean verbose, ActionListener<GetSnapshotsResponse> listener) {
+        List<Future<List<SnapshotInfo>>> futures = new ArrayList<>(repos.size());
 
-            final RepositoryData repositoryData;
-            if (isCurrentSnapshotsOnly(request.snapshots()) == false) {
-                repositoryData = snapshotsService.getRepositoryData(repository);
-                for (SnapshotId snapshotId : repositoryData.getAllSnapshotIds()) {
-                    allSnapshotIds.put(snapshotId.getName(), snapshotId);
+        // run concurrently for all repos on SNAPSHOT thread pool
+        for (final RepositoryMetaData repo : repos) {
+            futures.add(threadPool.executor(ThreadPool.Names.SNAPSHOT).submit(
+                    () -> getSingleRepoSnapshotInfo(repo.name(), snapshots, ignoreUnavailable, verbose)));
+        }
+        assert repos.size() == futures.size();
+
+        Map<String, List<SnapshotInfo>> successfulResponses = new HashMap<>();
+        Map<String, ElasticsearchException> failedResponses = new HashMap<>();
+
+        for (int i = 0; i < repos.size(); i++) {
+            final String repo = repos.get(i).name();
```

Comment:
I've implemented it here https://github.com/elastic/elasticsearch/pull/42090/commits/94e60ddb60c0ccf551ff78f498e80dc260e14297. However, it looks more cumbersome than the previous version for two reasons:
1) There is not repoName in `SnapshotInfo`, so we need to put this name into the tuple, because `GroupedActionListener` does not maintain the order of the results, so it's not possible to read this name from the list of repos.
2) There is no `Either`/`Xor` in Java to pack successful xor unsuccessful result, so we use Tuple with nullable values instead.  
This results in very long type signatures. I prefer to rollback this change.

Reply:
Actually, by making `GetSnapshotsResponse.Response` public both problems are resolved and code looks much better. So let's keep this change. https://github.com/elastic/elasticsearch/pull/42090/commits/c00640fba25712c25cbe45cd96eaad41f224f738